### PR TITLE
SLCORE-187 Track sever issues in connected mode in language server

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

File: language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+import static org.sonarsource.sonarlint.core.client.api.util.FileUtils.toSonarQubePath;
+
+class ServerIssueTracker {
+
+  private final ConnectedSonarLintEngine engine;
+  private final ServerConfiguration serverConfiguration;
+  private final String moduleKey;
+  private final Path baseDir;
+  private final Logger logger;
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java]: Code Smell Code Smell: Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". ([squid:S1312](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1312))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=195&issues=AWNdRMDonxz_298731lo&open=AWNdRMDonxz_298731lo)

[ISSUE_KEY]: AWNdRMDonxz_298731lo

File: language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  ServerIssueTracker(ConnectedSonarLintEngine engine, ServerConfiguration serverConfiguration,
+    String moduleKey, Path baseDir, org.sonarsource.sonarlint.core.tracking.Logger logger) {
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java]: Code Smell Code Smell: Replace this fully qualified name with "Logger" ([squid:S1942](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1942))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=195&issues=AWNdRMDonxz_298731lp&open=AWNdRMDonxz_298731lp)

[ISSUE_KEY]: AWNdRMDonxz_298731lp

File: language-server/src/test/java/org/sonarlint/languageserver/ServerIssueTrackerTest.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    return newTracker(baseDir, engine);
+  }
+
+  // create uniquely identifiable issue
+  private Issue mockIssue() {
+    Issue issue = mock(Issue.class);
+
+    // basic setup to prevent NPEs
+    when(issue.getInputFile()).thenReturn(mock(ClientInputFile.class));
+    when(issue.getMessage()).thenReturn("dummy message " + counter.incrementAndGet());
```
Comment (dbmeneses) [inline language-server/src/test/java/org/sonarlint/languageserver/ServerIssueTrackerTest.java]: AtomicInteger is typically for concurrent operations. Why not just do `++i`?

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 4]
```diff
@@ -650,24 +652,60 @@ public AnalysisResultsWrapper analyze(URI uri, String content, IssueListener iss
...
 
       return new AnalysisResultsWrapper(analysisResults, analysisTime);
     }
+
+    private AnalysisResults analyze(ConnectedAnalysisConfiguration configuration, IssueCollector collector) {
+      return engine.analyze(configuration, collector, logOutput, null);
+    }
+  }
+
+  static class IssueCollector implements IssueListener {
```
Comment (dbmeneses) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Maybe could be a lambda that collects to a local list?

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 5]
```diff
@@ -650,24 +652,60 @@ public AnalysisResultsWrapper analyze(URI uri, String content, IssueListener iss
...
+    final List<Issue> issues = new LinkedList<>();
+
+    @Override
+    public void handle(Issue issue) {
+      issues.add(issue);
+    }
+  }
+
+  private class ServerIssueTrackingLogger implements org.sonarsource.sonarlint.core.tracking.Logger {
```
Comment (dbmeneses) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: It doesn't need to be a inner class and I would rename it to something such as `IssueTrackingLoggerAdapter`.

File: language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    engine.downloadServerIssues(serverConfiguration, moduleKey);
+
+    Collection<Issue> issuesWithFile = issues.stream().filter(issue -> issue.getInputFile() != null).collect(Collectors.toList());
+    Collection<String> relativePaths = getRelativePaths(baseDir, issuesWithFile);
+    Map<String, List<Trackable>> trackablesPerFile = getTrackablesPerFile(baseDir, issuesWithFile);
+    IssueTrackerCache cache = createCurrentIssueTrackerCache(relativePaths, trackablesPerFile);
+
+    return getCurrentTrackables(relativePaths, cache)
+      .stream()
+      .map(trackable -> new DelegatingIssue(trackable.getIssue()) {
```
Comment (dbmeneses) [inline language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java]: why not do this in `getCurrentTrackables` and avoid creating a list twice?

File: language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+import static org.sonarsource.sonarlint.core.client.api.util.FileUtils.toSonarQubePath;
+
+class ServerIssueTracker {
+
+  private final ConnectedSonarLintEngine engine;
+  private final ServerConfiguration serverConfiguration;
+  private final String moduleKey;
+  private final Path baseDir;
+  private final Logger logger;
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/ServerIssueTracker.java]: Code Smell Code Smell: Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". ([squid:S1312](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1312))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=195&issues=AWNdRMDonxz_298731lo&open=AWNdRMDonxz_298731lo)

[ISSUE_KEY]: AWNdRMDonxz_298731lo

File: language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java [hunk 8]
```diff
@@ -570,23 +778,36 @@ public WorkspaceService getWorkspaceService() {
...
+        if (args == null) {
+          break;
+        }
         if (args.size() != 1) {
-          warn("Expecting 1 argument");
+          logger.warn("Expecting 1 argument");
         } else {
           String ruleKey = parseToString(args.get(0));
-          RuleDetails ruleDetails = engine.getRuleDetails(ruleKey);
+          // TODO use the correct engine (currently only Atom reaches this code, and it doesn't have connected mode, yet)
```
Comment (SonarTech) [inline language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=195&issues=AWNdRMBQnxz_298731ln&open=AWNdRMBQnxz_298731ln)

[ISSUE_KEY]: AWNdRMBQnxz_298731ln

**Pull request merged**